aspnetcore/mvc/models/file-uploads.md

```diff
@@ -1,195 +1,656 @@
+> * <xref:System.IO.Path.GetTempFileName*>
 
-Files uploaded using the `IFormFile` technique are buffered in memory or on disk on the web server before being processed. Inside the action method, the `IFormFile` contents are accessible as a stream. In addition to the local file system, files can be streamed to [Azure Blob storage](/azure/visual-studio/vs-storage-aspnet5-getting-started-blobs) or [Entity Framework](/ef/core/index).
+### Upload small files with buffered model binding to a database
 
-To store binary file data in a database using Entity Framework, define a property of type `byte[]` on the entity:
+To store binary file data in a database using [Entity Framework](/ef/core/index), define a <xref:System.Byte> array property on the entity. The following example is similar to a scenario demonstrated in the sample app (*Upload one buffered file with one file upload control*):
 
 ```csharp
-public class ApplicationUser : IdentityUser
+public class AppFile
 {
-    public byte[] AvatarImage { get; set; }
+    public int Id { get; set; }
+    public byte[] Content { get; set; }
 }
 ```
 
-Specify a viewmodel property of type `IFormFile`:
+Specify a page model property for the class that includes an <xref:Microsoft.AspNetCore.Http.IFormFile>:
 
 ```csharp
-public class RegisterViewModel
+public class BufferedSingleFileUploadDbModel : PageModel
 {
-    // other properties omitted
+    ...
+
+    [BindProperty]
+    public BufferedSingleFileUploadDb FileUpload { get; set; }
+
+    ...
+}
 
-    public IFormFile AvatarImage { get; set; }
+public class BufferedSingleFileUploadDb
+{
+    [Required]
+    [Display(Name="File")]
+    public IFormFile FormFile { get; set; }
 }
 ```
 
 > [!NOTE]
-> `IFormFile` can be used directly as an action method parameter or as a viewmodel property, as shown above.
+> <xref:Microsoft.AspNetCore.Http.IFormFile> can be used directly as an action method parameter or as a bound model property. The prior example uses a bound model property.
+
+The `FileUpload` is used in the Razor Pages form:
+
+```cshtml
+<form enctype="multipart/form-data" method="post">
+    <dl>
+        <dt>
+            <label asp-for="FileUpload.FormFile"></label>
+        </dt>
+        <dd>
+            <input asp-for="FileUpload.FormFile" type="file">
+        </dd>
+    </dl>
+    <input asp-page-handler="Upload" class="btn" type="submit" value="Upload">
+</form>
+```
 
-Copy the `IFormFile` to a stream and save it to the byte array:
+When the form is POSTed to the server, copy the <xref:Microsoft.AspNetCore.Http.IFormFile> to a stream and save it as a byte array in the database:
 
 ```csharp
-// POST: /Account/Register
-[HttpPost]
-[AllowAnonymous]
-[ValidateAntiForgeryToken]
-public async Task<IActionResult> Register(RegisterViewModel model)
+public async Task<IActionResult> OnPostUploadAsync()
 {
-    ViewData["ReturnUrl"] = returnUrl;
-    if  (ModelState.IsValid)
+    using (var memoryStream = new MemoryStream())
```

Comment:
`using var memoryStream = new MemoryStream()` for 💯 points.

Reply:
Also, should this one have a line that validates file size before doing the `ToArray()`? Even in the ordinary case it's a DOS vector since the server is allocating 2x the memory for every 1x file the client uploads.

Reply:
> `using var memoryStream = new MemoryStream()`

That's a C# 8.0 thing, right? Doesn't it break in a 2.2 app without setting the lang version up?